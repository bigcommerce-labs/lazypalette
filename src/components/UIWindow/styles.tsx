import styled, { keyframes } from 'styled-components';

export const Content = styled.div`
    margin: 0;
    padding: .5rem .75rem;
    overflow: auto;
`;

export const Header = styled.div`
    align-items: center;
    background-color: ${({ theme }) => theme.colors.background};
    cursor: pointer;
    cursor: move;
    display: flex;
    justify-content: flex-end;
    height: 1rem;
    padding: 0.75rem 0.75rem 0.75rem 1.5rem;
    margin: 0;
    font-weight: ${({ theme }) => theme.typography.fontWeight.normal};
    color: ${({ theme }) => theme.colors.primaryText};
`;

Header.defaultProps = {
    theme: {
        colors: {
            primaryText: '#FFFFFF',
        },
        typography: {
            fontWeight: {
                bold: 200,
            },
        },
    },
};

Header.displayName = 'Header';

export const Title = styled.h4`
    margin: 0;
    font-weight: ${({ theme }) => theme.typography.fontWeight.normal};
    color: ${({ theme }) => theme.colors.primaryText};
`;

Title.defaultProps = {
    theme: {
        colors: {
            primaryText: '#FFFFFF',
        },
        typography: {
            fontWeight: {
                normal: 200,
            },
        },
    },
};

interface StyledWindowProps {
    position?: { x: number, y: number };
}

const openAnimation = keyframes`
  0% {
    opacity: 0;
    transform: scale(0.85);
  }

  100% {
    opacity: 1;
    transform: scale(1);
  }
`;

export const StyledWindow = styled.div.attrs<StyledWindowProps>({
    style: ({ position }: StyledWindowProps) => ({
        left: position ? `${position.x}px` : undefined,
        top: position ? `${position.y}px` : undefined,
    }),
})`
    position: absolute;
    background-color: ${({ theme }) => theme.colors.empty};
    border-radius: 6px;
    box-shadow: ${({ theme }) => theme.elevation.floating};
    overflow: auto;
    animation: ${openAnimation} 50ms linear;
    pointer-events: auto;
    z-index: ${({ theme }) => theme.layers.higher};
`;

StyledWindow.defaultProps = {
    theme: {
        colors: {
            empty: '#FFFFFF',
            guideText: '#AAAAAA',
        },
        elevation: {
            floating: '0 2px 12px rgba(48, 53, 64, 0.2)',
        },
        layers: {
            higher: 100,
        },
    },
};

export const Overlay = styled.div`
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    overflow: hidden;
`;

export const NavItem = styled.div`
    text-align: center;
    text-decoration: none;
    text-shadow: 0 1px 0 ${({ theme }) => theme.colors.background};
    font-size: ${({ theme }) => theme.typography.fontSize.smaller};

    :after {
        content: 'âœ•';
        color: ${({ theme }) => theme.colors.primaryText};

        :visited, :hover, :active {
            color: ${({ theme }) => theme.colors.primaryText};
        }
    }

    :hover {
      cursor: pointer;
    }
`;

NavItem.displayName = 'NavItem';

NavItem.defaultProps = {
    theme: {
        colors: {
            background: '#F5F7FA',
            empty: '#FFFFFF',
            guideText: '#AAAAAA',
        },
        typography: {
            fontSize: {
                smaller: '14px',
            },
        },
    },
};
