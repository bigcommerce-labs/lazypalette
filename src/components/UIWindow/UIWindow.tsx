import React, { MouseEvent, PureComponent } from 'react';

import { Content, Header, StyledWindow } from './styles';

interface UIWindowProps {
    children: any;
    id: string;
    position: { x: number, y: number };
    title?: string;
    topmost: boolean;
    onClose(id: string): void;
}

class UIWindow extends PureComponent<UIWindowProps> {
    wrapperRef: any;

    constructor(props: UIWindowProps) {
        super(props);
        this.wrapperRef = React.createRef();
    }

    componentDidMount() {
        document.addEventListener('mousedown', this.handleClickOutside);
    }

    componentWillUnmount() {
        document.removeEventListener('mousedown', this.handleClickOutside);
    }

    handleClickOutside: EventListener = event => {
        if (!this.props.topmost) {
            return;
        }

        if (this.wrapperRef && this.wrapperRef.current && !this.wrapperRef.current.contains(event.target)) {
            if (this.props.onClose) {
                this.props.onClose(this.props.id);
            }
        }
    };

    handleModalClick = (event: MouseEvent<HTMLDivElement>) => {
        event.stopPropagation();
    };

    render() {
        const {
            children,
            position,
            title,
        } = this.props;

        return (
            <div ref={this.wrapperRef}>
                <StyledWindow
                    onClick={this.handleModalClick}
                    position={position}
                >
                    {title &&
                        <Header>{title}</Header>}
                    <Content>{children}</Content>
                </StyledWindow>
            </div>
        );
    }
}

export default UIWindow;
