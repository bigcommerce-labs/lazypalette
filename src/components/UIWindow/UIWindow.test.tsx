import { mount, shallow } from 'enzyme';
import React from 'react';

import UIWindow from './UIWindow';

const mockCloseWindow = jest.fn();
const addEventListener = jest.fn();
const removeEventListener = jest.fn();

document.addEventListener = addEventListener;
document.removeEventListener = removeEventListener;

describe('UIWindow', () => {
    beforeEach(() => {
        mockCloseWindow.mockReset();
        addEventListener.mockReset(); //tslint:disable-line
        removeEventListener.mockReset(); //tslint:disable-line
    });

    describe('when given a window', () => {
        it('renders correctly', () => {
            const uiWindow = shallow(<UIWindow
                id="1234"
                position={{ x: 5, y: 10}}
                title="hello"
                topmost={true}
                onClose={mockCloseWindow}>
                <p>blah</p>
            </UIWindow>);

            expect(uiWindow).toMatchSnapshot();
        });
    });

    describe('when mounting', () => {
        it('adds an event listener', () => {
            shallow(<UIWindow
                id="1234"
                position={{ x: 5, y: 10}}
                title="hello"
                topmost={true}
                onClose={mockCloseWindow}>
                <p>blah</p>
            </UIWindow>);

            expect(addEventListener).toHaveBeenCalled();
            expect(removeEventListener).not.toHaveBeenCalled();
        });
    });

    describe('when unmounting', () => {
        it('removes an event listener', () => {
            const uiWindow = shallow(<UIWindow
                id="1234"
                position={{ x: 5, y: 10}}
                title="hello"
                topmost={true}
                onClose={mockCloseWindow}>
                <p>blah</p>
            </UIWindow>);

            removeEventListener.mockReset();
            addEventListener.mockReset();
            uiWindow.unmount();
            expect(removeEventListener).toHaveBeenCalled();
            expect(addEventListener).not.toHaveBeenCalled();
        });
    });

    describe('when a click occurs', () => {
        describe('when the window is topmost', () => {
            let uiWindow: any;

            beforeEach(() => {
                uiWindow = mount(<UIWindow
                    id="1234"
                    position={{ x: 5, y: 10 }}
                    title="hello"
                    topmost={true}
                    onClose={mockCloseWindow}
                    windowRef={React.createRef()}>
                    <p id="blah">blah</p>
                </UIWindow>);
            });

            describe('inside the modal', () => {
                it('does not call the onClose prop unless on X', () => {
                    const paragraph = uiWindow.find('#blah').first().getDOMNode();
                    uiWindow.instance().handleDocumentClick({ target: paragraph, button: 0 });

                    expect(mockCloseWindow).not.toHaveBeenCalled();
                });

                it('does call the onClose prop when clicking X', () => {
                    const iconX = uiWindow.find('NavItem');
                    iconX.simulate('click');
                    expect(mockCloseWindow).toHaveBeenCalled();
                });
            });

            describe('outside the modal', () => {
                it('does call the onClose prop', () => {
                    const paragraph = mount(<p>blah</p>).first().getDOMNode();
                    uiWindow.instance().handleDocumentClick({ target: paragraph, button: 0 });

                    expect(mockCloseWindow).toHaveBeenCalledWith('1234');
                });

            });
        });

        describe('when the window is not topmost', () => {
            let uiWindow: any;

            beforeEach(() => {
                uiWindow = mount(<UIWindow
                    id="1234"
                    position={{ x: 5, y: 10 }}
                    title="hello"
                    topmost={false}
                    onClose={mockCloseWindow}
                    windowRef={React.createRef()}>
                    <p id="blah">blah</p>
                </UIWindow>);
            });

            describe('outside the modal', () => {
                it('does not call the onClose prop', () => {
                    const paragraph = mount(<p>blah</p>).first().getDOMNode();
                    uiWindow.instance().handleDocumentClick({ target: paragraph, button: 0 });

                    expect(mockCloseWindow).not.toHaveBeenCalled();
                });
            });
        });
    });
});
