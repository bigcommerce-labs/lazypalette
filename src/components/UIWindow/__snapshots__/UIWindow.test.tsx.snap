// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIWindow when given a window renders correctly 1`] = `
<div>
  <styled.div
    onClick={[Function]}
    position={
      Object {
        "x": 5,
        "y": 10,
      }
    }
    theme={
      Object {
        "colors": Object {
          "empty": "#FFFFFF",
          "guideText": "#AAAAAA",
        },
        "elevation": Object {
          "raised": 1,
        },
      }
    }
  >
    <styled.h3
      theme={
        Object {
          "colors": Object {
            "primaryText": "#FFFFFF",
          },
          "typography": Object {
            "fontWeight": Object {
              "bold": 200,
            },
          },
        }
      }
    >
      hello
    </styled.h3>
    <styled.div>
      <p>
        blah
      </p>
    </styled.div>
  </styled.div>
</div>
`;
